app/ml_models/rnn/generate.py

```diff
@@ -1,54 +1,74 @@
 import torch
 from app.ml_models.rnn.helpers import get_input_tensor
 from app.ml_models.rnn.rnn_model import RNN
+from random import choice as choose
 
 
-def generate_word(
-    model: RNN,
-    N_LETTERS: int,
-    ALL_LETTERS: set,
-    start_letter: str = "a",
-    maxn: int = 20,
-    temp: float = 0,
-) -> str:
-    """Generate a new Slang word.
+def next_char(out, temperature):
+    # Softmax of the last dimension
+    if torch.distributions.Uniform(0, 1).sample() < temperature:
+        probs = torch.softmax((out), -1)
+        #probs = torch.softmax(temperature*(out), -1) # This is good for randomness (temperature < 1)
+        choice = torch.multinomial(probs.squeeze(0), 1)
+    else:
+        choice = torch.argmax(out, dim=2)
+    return choice
+
+
+def generate_word(model, dataset, start_letter=None,  max_len=20, temperature=0.25, device='cpu'):
+    """Generate a new word.
 
     Args:
         model: Pre-trained Recurrent Neural Network model.
-        N_LETTERS: Number of unique letters found in the training corpus.
-        ALL_LETTERS: All letters found in the training data.
+        dataset: WordLevelDataset object 
         start_letter: Letter to start the word with.
-        maxn: Maximum number of letters to be used.
+        max_len: Maximum number of letters to be used.
         temp: Temperature used for sampling.
+        device: torch device string
     """
+    # Evaluation mode
+    model.eval()
+    # no gradient
     with torch.no_grad():
-        input = get_input_tensor(
-            start_letter,
-            N_LETTERS=N_LETTERS,
-            ALL_LETTERS=ALL_LETTERS,
-        )
-        hidden = model.initHidden()
-        output_name = start_letter
-        for i in range(maxn):
-            output, hidden = model(input[0], hidden)
-            if torch.distributions.Uniform(0, 1).sample() < temp:
-                probs = torch.softmax(output, 1)
-                dist = torch.distributions.Categorical(probs)
-                pick = dist.sample()
-            else:
-                topv, topi = output.topk(1)
-                pick = topi[0][0]
-            if pick == N_LETTERS - 1:
-                break
-            else:
-                letter = ALL_LETTERS[pick]
-                output_name += letter
-            input = get_input_tensor(
-                letter,
-                N_LETTERS=N_LETTERS,
-                ALL_LETTERS=ALL_LETTERS)
-
-        return output_name
+        # Hidden stuff initialized to None (pytorch makes this zeros automatically)
+        h = None
+        it = 0
+
+        # Always generate the Beginning of Word token first and feed it to the RNN
+        # TODO: Maybe don't ? It results in a lot of copying behaviour for small datasets
+        # idxs = torch.Tensor([dataset.char_to_idx_dict["<BOS>"]]
+        #                     ).long().unsqueeze(0).to(device)
+        # out, h = model(idxs, h)
+
+        choice = torch.Tensor([-99])
```

Comment:
Does the -99 refer to something specific to the vocabulary or is it just a random value to initialize this with? If the former, we may wish to change this to something more generic.

Reply:
This is just random, it gets overwritten immediately if all goes well. don't know what the best arbitrary value to choose is, but indices for embeddings can't be negative afaik so any negative number seems like a safe choice